src/lib/HST/Frontend/Syntax.hs

```diff
@@ -47,41 +55,42 @@ newtype Module a = Module [Decl a]
 deriving instance EqAST a => Eq (Module a)
 deriving instance ShowAST a => Show (Module a)
 
+-------------------------------------------------------------------------------
+-- Declarations                                                              --
+-------------------------------------------------------------------------------
+
 -- | A declaration.
 --
 --   Data declarations only contain the information relevant for the pattern
 --   matching compiler and should not be transformed back.
 --
 --   The only supported kind of pattern bindings, variable patterns, are
 --   represented by function bindings.
-data Decl a = DataDecl (DeclHead a) [ConDecl a]
-            | TypeSig (SrcSpan a) [Name a] (TypeExp a)
+data Decl a = DataDecl (Name a) [ConDecl a]
             | FunBind (SrcSpan a) [Match a]
```

Comment:
The reason I implemented the `TypeSig` constructor was to support and retain type signatures in `where` bindings, as these are quite common. In my opinion, this is reason enough to leave the `TypeSig` constructor inserted.
Or do you have a different idea for those type signatures?

Reply:
Ah okay, that makes sense. I didn't think about `where` clauses and `let` expressions. However, as far as I know `where` clauses are not supported anyway and will even be discarded silently at the moment. Thus, we should probably remove them from the intermediate syntax and print an error message if there is a `where` clause.
In `let` expressions it would still be nice to preserve type signatures. Maybe we can add a more general `UnsupportedDecl` constructor for this purpose. This constructor would also allow us to remove the input module parameter from `ToHSE.transformModule`.

Reply:
See 60574d0f9ba616824bb6b22a7cf7475a8a169495.